src/haz3lcore/Animation.re

```diff
@@ -0,0 +1,232 @@
+open Util;
+open Js_of_ocaml;
+
+/* This implements arbitrary gpu-accelerated css position
+ * and scale transition animations via the the FLIP technique
+ * (https://aerotwist.com/blog/flip-your-animations/).
+ *
+ * From the client perspective, it suffices to call the request
+ * method with a list of the DOM element ids to animate, as well
+ * as some animation settings (keyframes, duration, easing).
+ *
+ * Some common keyframes are provided in the module at the bottom */
+
+/* This is an extremely partial implementation of the Web Animations
+ * API, which currently does not have Js_of_ocaml wrappers */
+module Js = {
+  /* CSS property-value pairs */
+  type keyframe = (string, string);
+
+  type options = {
+    duration: int,
+    easing: string,
+  };
+
+  /* Options for CSS Animations API animate method */
+  type animation = {
+    options,
+    keyframes: list(keyframe),
+  };
+
+  /* Position & dimensions for a DOM element */
+  type box = {
+    top: float,
+    left: float,
+    height: float,
+    width: float,
+  };
+
+  let box_of = (elem: Js.t(Dom_html.element)): box => {
+    let container_rect = elem##getBoundingClientRect;
+    {
+      top: container_rect##.top,
+      left: container_rect##.left,
+      height: Js.Optdef.get(container_rect##.height, _ => (-1.0)),
+      width: Js.Optdef.get(container_rect##.width, _ => (-1.0)),
```

Comment:
What's the deal with the defaults being -1.0 for these and 0.0 for the ones below?